up42/jobcollection.py

```diff
@@ -33,10 +33,69 @@ def __repr__(self):
             f"auth={self.auth})"
         )
 
-    # TODO: Make class subscriptable.
+    def __getitem__(self, index: int) -> Job:
+        return self.jobs[index]
+
+    def __iter__(self):
+        for job in self.jobs:
+            yield job
+
+    def apply(
+        self, worker: Callable, only_succeeded: bool = True, **kwargs
+    ) -> Dict[str, Any]:
+        """
+        Helper function to apply `worker` on all jobs in the collection.
+        `worker` needs to accept `Job` as first argument. For example, a
+        lambda function that returns the job info:
+        ```python
+        self.apply(lambda job: job._get_info())
+        ```
+
+        Args:
+            worker: A function to apply on all jobs in the collection.
+            only_succeeded: Only apply to succeeded jobs (default is `True`).
+            kwargs: additional keyword arguments to pass to `worker`.
+        Returns:
+            Dictionary where the key is the job id and the value the return
+            of `worker`.
+        """
+        out_dict = {}
+        for job in self.jobs:
+            if only_succeeded:
+                if job.is_succeeded:
+                    out_dict[job.job_id] = worker(job, **kwargs)
+            else:
+                out_dict[job.job_id] = worker(job, **kwargs)
+        if not out_dict:
+            if self.jobs:
+                raise ValueError(
+                    "All jobs have failed! Cannot apply over an empty succeeded job list."
+                )
+            if not self.jobs:
```

Comment:
Nice functionality. A bit simpler like this maybe, at least the raises top&bottom.
```
if not self.jobs:
    raise ValueError("This is an empty JobCollection. Cannot apply over an empty job list.")

out_dict = {}
for job in self.jobs:
     if only_succeeded:
            if not job.is_succeeded:
                 continue
     out_dict[job.job_id] = worker(job, **kwargs)

if not out_dict:
   raise ValueError("All jobs have failed! Cannot apply over an empty succeeded job list.")
```

Reply:
Very good suggestion, looks much better, applied!